### ISSUE TITLE
creating a broken sponsorship makes it impossible to manage them

### ISSUE THREAD

**Issue by kristoff-it** [forem/forem#14386]
I created a sponsorship where I filled the last three fields with bogus values (`?a`, `?b`, `?c`) and now accessing the sponsorships admin page causes a rails error.

Here's the relevant part in the Honeybadger report.
![image](https://user-images.githubusercontent.com/1950733/127509125-ede69f7b-b4f6-4ea4-a412-4282417dd333.png)

This is a problem also because I can't disable the sponsorship. 
Any suggestion on how to quickly disable it at least?

**Comment by djuber** [comment]
I can reproduce this locally following the suggested input.

```
NameError: wrong constant name ?c
from /home/djuber/src/forem/vendor/cache/ruby/2.7.0/gems/activesupport-6.1.4/lib/active_support/inflector/methods.rb:274:in `const_get'
```

The issue is the "Sponsorable Type" is interpreted in rails as a class name (in DEV they're all set to `ActsAsTaggableOn::Tag` or are blank) - this should almost certainly have been an enumerated list of possible class names, rather than a string input that permitted creating an unloadable object (actually, the Sponsorship is able to be initialized, the `:sponsorable` relation that's being included is the problem). The admin side of the software probably has a lot more "sharp corners" than it should.

I did not see anything in the api that would permit deleting this outside of the (broken) web ui. 

If you've done this on your own instance the fastest way to clear it would be via the rails console 

```
Sponsorship.where(sponsorable_type: '?c').destroy_all
```

If you have database access it would be effectively the same `DELETE FROM sponsorships WHERE sponsorable_type = '?c';`  

If  you've done this on a Forem hosted instance, someone will need to do this for you I think (I checked the DEV.to database and could not see this, and didn't see it in our honeybadger logs so I suspect this is self-hosted or a different server).

**Comment by kristoff-it** [comment]
Yes, it's a self-hosted instance, I'm experimenting with all these features before opening it to the public. 
Thank you for the thorough reply, the shell cleanup worked flawlessly.

**Comment by djuber** [comment]
I'd like to keep this open as a bug report (it shouldn't have been quite so easy for you to get into that situation), but pivot to preventing saving these "broken" sponsorships rather than trying to load/display/manage them after they're saved.

I believe this means one or both of these could be added to the software:

- Sponsorship should validate that sponsorable_type is a valid constant before permitting saving (feels like a small addition)

- The admin UI for sponsorship creation/editing should have a dropdown of permitted types. Tag seems to be the only common type on DEV, it's unclear what else could be sponsored and might be an editorial decision to make that choice, feels larger than the "type is a valid model name" first point

**Comment by mikeydorje** [comment]
Code review suggests this may be resolved or no longer applicable due to changes in this area since the issue was filed. Closing due to age and low activity. Please reopen with current repro steps if this persists.
